core/pta/hwrng.c

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2018, Linaro Limited
+ * Copyright (c) 2021, EPAM Systems. All rights reserved.
+ *
+ * Based on plat-synquacer/rng_pta.c
+ *
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <rng_support.h>
+#include <rng_pta_client.h>
+
+#define PTA_NAME "rng.pta"
+
+static TEE_Result rng_get_entropy(uint32_t types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint8_t *e = NULL;
+	uint32_t i = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	e = (uint8_t *)params[0].memref.buffer;
+	if (!e)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < params[0].memref.size; i++)
+		e[i] = hw_get_random_byte();
```

Comment:
Hi,
Reviewing this commit, I'm surprised about the usage of the direct call to `hw_get_random_byte()` which, I think,  must not be exposed at that level. In the entire code, the interface that might centralized the call to rng is through the crypto interface `crypto_rng_read()`.

I understand that the PTA interface must only be available while using `RNG_HW`, in that case, why not defining the PTA interface linked to `RNG_HW` defined.
The access to the RNG can be done from TA or PTA using the same crypto interface rather than using `get_hw_random_byte()` that **_MUST_** be implemented in a low level driver as used in a `__weak__` framework function.

Using a single interface will allow to control access to the RNG whatever the implementation and whatever the source as the is only one way to get it.
BR,

Reply:
> Reviewing this commit, I'm surprised about the usage of the direct call to `hw_get_random_byte()` which, I think, must not be exposed at that level.

What is the problem with `hw_get_random_byte()`? Except that one could wish for a function allowing better efficiency, but that another issue.

Reply:
Question is more why do we need to call `hw_get_random_byte()` as we have a `crypto_rng_read()` function that is the only one calling the `hw_get_random_byte()` from the hw_rng.c ?

Why not directly use the `crypto_rng_read()` here?

Reply:
> I understand that the PTA interface must only be available while using RNG_HW, in that case, why not defining the PTA interface linked to RNG_HW defined.

The idea of using `hw_get_random_byte()` specifically is _not_ to let it work with PRNG in any circumstances _and_ keep it simple.
Surely `crypto_rng_read()` can be used here, but in that case we need to add several extra build checks to cover more configuration states (check for condition `CFG_WITH_SOFTWARE_PRNG=n`), which adds some complexity for no really good reason.
Also, should `CFG_WITH_SOFTWARE_PRNG` ever be modified we would have many different places that need to be remembered and taken care of (`core/crypto/sub.mk` and `mk/config.mk` + platform `conf.mk`), so that hwrng PTA does not ever accidentaly use PRNG.

Long story short, it is implemented that way to be more simpler and safer. However I don't know exact purpose of that interface, so if I misused it I'll gladly change it for `crypto_rng_read()`.

Reply:
OK, maybe later discussion around the cryto_rng_interface.

I would say that a simple 
`ifeq ($(CFG_WITH_SOFTWARE_PRNG),n)`
`srcs-$(CFG_HWRNG_PTA) += hwrng.c`
`endif`
probably solve this issue and do not introduce too much complexity?
BR

Reply:
> `ifeq ($(CFG_WITH_SOFTWARE_PRNG),n)`
> `srcs-$(CFG_HWRNG_PTA) += hwrng.c`
> `endif`

I've just removed such code by @jforissier request, see [this](https://github.com/OP-TEE/optee_os/pull/4641#discussion_r639561687) comment.
Probably `mk/config.mk` is the place.

Reply:
done, see eddb770 & d9bd599